File: 509-FibonacciNumber/509-FibonacciNumber.cpp
// Last updated: 8/2/2026, 6:54:26 PM
1class Solution {
2public:
3const int INF = 1e9;
4    int solve(int amount,int i,vector<vector<int>> &dp, vector<int>& coins)
5    {
6        if(amount == 0)
7        {
8            return 0;
9        }
10        if(i==coins.size())
11        {
12            return INF;
13        }
14        if(dp[i][amount]!=-1)
15        {
16            return dp[i][amount];
17        }
18        int take = INF;
19        if(coins[i]<=amount)
20        {
21            take = 1 + solve(amount-coins[i],i,dp,coins);
22        }
23        int notTake = solve(amount,i+1,dp,coins);
24        dp[i][amount] = min(take,notTake);
25        return dp[i][amount];
26    }
27    int coinChange(vector<int>& coins, int amount) {
28        vector<vector<int>> dp(coins.size(),vector<int>(amount+1,-1));
29        int ans = solve(amount,0,dp,coins);
30        return (ans==INF)?-1:ans;
31    }
32};

File: 509-FibonacciNumber/509-FibonacciNumber.py
# Last updated: 9/20/2026, 8:17:50 PM
1class Solution:
2
3    def solve(self, dp, n):
4        if n == 0 or n == 1:
5            return n
6
7        if dp[n] != -1:
8            return dp[n]
9
10        dp[n] = self.solve(dp, n-1) + self.solve(dp, n-2)
11        return dp[n]
12
13    def fib(self, n):
14        dp = [-1] * (n+1)
15        return self.solve(dp, n)